### DISCUSSION TITLE
Atom Store

### DISCUSSION THREAD

**Discussion by meatballs** [anvilistas/anvil-labs#9] [Ideas]
How about defining an interface (possibly an ABC) for a store?

With methods to save an atom and retrieve one by id.

We could then create concrete implementations of that interface for e.g. LoacalStorage, IndexedDb, data tables and users could create their own for whatever third party solution they want.

Rather than adding persistence methods to atoms, we delegate that job to a separate object.

**Comment by meatballs** [comment]
We could even have multiple data tables schema, each with its own Store class.

**Reply by s-cork** [reply]
Yeah I like the sound of this. 

So far I've implemented two variations of updating a db or local storage. 

The first is to use a render method. Any time something I'm interested in changes the render method gets called and does the update. 

The other was to label actions decorators with certain keywords. 
`@action(server_call="foo")`. I could catch these actions in the final stage of the render cycle with the @subscribe decorator. 

Both were reasonable. But something more general purpose would be 👍. 

The other thing that I'm struggling with is to put anvil data table rows and search iterators into the mix in a clean way. I think when we get that right it's a game changer.
